### PR TITLE
feat: OpenAPI tool extracts POST operations

### PR DESCRIPTION
**Pull request by Cobalt0s**
# Description

This is an improvement to OpenAPI schema extractor tool.
Now it supports querying non-GET methods. In particular we are interested in POST which act as search in most APIs.

# Usage

Later dependent intercom PR benefits by adding `tickets` to metadata and Read can be done by POST. It wasn't present already because, there is no GET endpoint.

### CODE REVIEW

**Comment by Cobalt0s** [comment]
* **#1082** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1082?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1084** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1084?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1083** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1083?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Cobalt0s and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by Cobalt0s** [comment]
* **#1082** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1082?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1084** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1084?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1083** <a href="https://app.graphite.dev/github/pr/amp-labs/connectors/1083?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Cobalt0s and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: tools/fileconv/api3/queries.go [hunk 1]
```diff
@@ -15,20 +15,44 @@ type Explorer struct {
...
 	displayNameOverride map[string]string,
 	check ObjectCheck,
-) ([]Schema, error) {
-	schemas := make([]Schema, 0)
+) (Schemas, error) {
+	return e.ReadObjects("GET", ignoreEndpoints, true, objectEndpoints, displayNameOverride, check)
+}
+
+// SearchObjectsPost is the same as ReadObjectsGet but retrieves schemas for endpoints that perform reading via POST.
+func (e Explorer) SearchObjectsPost(
```
Comment (RajatPawar) [inline tools/fileconv/api3/queries.go:32]: As a convention, could this be `ReadObjectsPost`?

File: tools/fileconv/api3/queries.go [hunk 2]
```diff
@@ -15,20 +15,44 @@ type Explorer struct {
...
+	return e.ReadObjects("POST", fromEndpoints, false, objectEndpoints, displayNameOverride, check)
+}
+
+// ReadObjects has 2 reading modes.
+// One ignores the provided list of endpoints (ignore=true).
+// The other scopes the search to just the list of endpoints (ignore=false).
+func (e Explorer) ReadObjects(
+	operationName string,
+	endpoints []string,
+	ignore bool,
```
Comment (RajatPawar) [inline tools/fileconv/api3/queries.go:47]: Please call this `ignoreEndpoints` to be more readable

File: tools/fileconv/api3/queries.go [hunk 3]
```diff
@@ -15,20 +15,44 @@ type Explorer struct {
...
 	check ObjectCheck,
-) ([]Schema, error) {
-	schemas := make([]Schema, 0)
+) (Schemas, error) {
+	return e.ReadObjects("GET", ignoreEndpoints, true, objectEndpoints, displayNameOverride, check)
+}
+
+// SearchObjectsPost is the same as ReadObjectsGet but retrieves schemas for endpoints that perform reading via POST.
+func (e Explorer) SearchObjectsPost(
+	fromEndpoints []string,
```
Comment (RajatPawar) [inline tools/fileconv/api3/queries.go:33]: If this is the same as `ReadObjectsGet.ignoreEndpoints`, can you call it the same? `fromEndpoints` isn't very clear on what it does

File: tools/fileconv/api3/strategy.go [hunk 4]
```diff
@@ -6,23 +6,30 @@ import (
 	"github.com/amp-labs/connectors/common/handy"
 )
 
-// This strategy prunes URL Path that should be omitted during Schema extraction.
+// This strategy prunes URL Path that should be omitted or included during Schema extraction.
 // It allows hard coded endpoint Path, as well as simple star rules focusing on matching prefix or suffix.
+// First mode ignore endpoints and those matching suffixes, prefixes,
```
Comment (RajatPawar) [inline tools/fileconv/api3/strategy.go:11]:
```suggestion
// First mode ignores endpoints and those matching suffixes, prefixes,
```

File: tools/fileconv/api3/strategy.go [hunk 5]
```diff
@@ -6,23 +6,30 @@ import (
...
 type ignorePathStrategy struct {
-	ignoreEndpoints handy.StringSet
-	prefixes        []string
-	suffixes        []string
+	endpoints handy.StringSet
+	prefixes  []string
+	suffixes  []string
+	// ignore flag determines the mode.
+	// It changes how we answer questions: Do we "ignore", or do we "include" given endpoint?
+	ignore bool
```
Comment (RajatPawar) [inline tools/fileconv/api3/strategy.go:24]: This is hard to understand IMO. To make it clearer, can you split this into two strategies? One could be an ignoreObjects strategy, and one could be a `onlyObjects` strategy. They could be controlled via a flag respectively.

File: tools/fileconv/api3/queries.go [hunk 6]
```diff
@@ -15,20 +15,44 @@ type Explorer struct {
...
 
-// GetBasicReadObjects retrieves schemas that can be used by ListObjectMetadata.
+// ReadObjectsGet retrieves schemas that can be used by ListObjectMetadata.
 // objectEndpoints - optional map of Endpoint Path to ObjectName associated with it.
 // ignoreEndpoints - optional list of paths to ignore
 // displayNameOverride - optional map of ObjectName to DisplayName. This will override display name from OpenAPI doc.
-func (e Explorer) GetBasicReadObjects(
+func (e Explorer) ReadObjectsGet(
 	ignoreEndpoints []string,
 	objectEndpoints map[string]string,
```
[the remark this answers is not in the document]
Comment (Cobalt0s) [reply tools/fileconv/api3/queries.go:24]: Added detailed description to `ReadObjects` method which I reference from this doc.

File: tools/fileconv/api3/strategy.go [hunk 7]
```diff
@@ -6,23 +6,87 @@ import (
...
+
+// PathMatchingStrategy This resolver has 2 level checking.
+// Path matches if it is approved by allow resolver or denied if matches by deny resolver.
+// The tie of what takes precedence is resolved by priorityAllow flag.
+// If neither match we rely on Strict flag.
+// Strict flag will mark a path as ignored, use this when AllowPaths and DenyPaths are exhaustive lists.
+// By default, Strict is false meaning any path will match, unless specified.
+//
+// prunes URL Path that should be omitted or included during Schema extraction.
+type PathMatchingStrategy struct {
```
Comment (Cobalt0s) [inline tools/fileconv/api3/strategy.go:21]: @RajatPawar Intercom was the first deep connector to require POST operations in OpenAPI. The solution with ignoreEndpoints doesn't hold that strongly. I was thinking to create some kind of imporvement and this PR was just to make Intercom use schema from POST.

This `PathMatchingStrategy` can be passed by either `ReadObjectsGet` or `ReadObjectsPost`.
It gives wider control on what endpoints the tool should look at and which to ignore. The strict flag would indicate if the end user of a method has provided exhaustive list or not. 
Usually you must explore all endpoints when creating metadata for ListObjectMetadata.
For intercom search endpoints are all that matters for POST extraction, so the list is strict, "exhaustive" and ok to ingore non-mentioned paths.
image
image

**Review by Cobalt0s — COMMENTED** [review]
@RajatPawar, I have added better strategy for Path matching and some unit tests, which should explain and enforce such functionality.

**Review by RajatPawar — APPROVED** [review]

**Pull request merged**